# Classification - Mini-Project 2

***Edit this cell with your name(s), tutorial number(s) and ID(s)***

---

Name: Malak Saleh

ID: 58-15339

Tutorial: 8

---

Name: Jana Salama

ID: 58-4500

Tutorial: 8

---


## Dataset Description

The following dataset includes information about loan applicants, including factors such as income, credit score, employment status, loan details, and other indicators of financial stability, along with the final decision showing whether each loan was approved or not.

| Column | Description|
|-|-|
|ApplicationNumber|Unique identifier assigned to each loan application|
|Age|Applicant’s age in years|
|AnnualIncome|Applicant’s yearly income|
|CreditScore|A score representing the applicant’s creditworthiness|
|EmploymentStatus|Applicant’s current employment situation (Employed, Unemployed, Self-Employed)|
|EducationLevel|Highest educational qualification attained (Highschool, Bachelor, Master, Doctorate, Diploma)|
|LoanAmount|Total amount of money requested for the loan|
|LoanDuration|Duration of the loan in months|
|MaritalStatus|Applicant’s marital state (Divorced, Married, Single, Widowed)|
|NumberOfDependents|Number of individuals financially dependent on the applicant|
|HomeOwnershipStatus|Applicant’s housing status (Mortagage, Own, Rent, Other)|
|BankruptcyHistory|Indicates whether the applicant has previously declared bankruptcy (0 = No, 1 = Yes)|
|LoanPurpose|The primary reason for taking the loan (Debt Consolidation, Home Improvement, Education, Personal)|
|PreviousLoanDefaults|Indicates if the applicant has defaulted on any previous loans (0 = No, 1 = Yes)|
|MonthlyLoanPayment|Amount the applicant would need to pay monthly to repay the loan|
|MonthlyIncome|Average monthly income of the applicant|
|JobTenure|Number of years the applicant has been in their current job|
|LoanApproved|Indicates loan approval status (No = Not Approved, Yes = Approved)|

## Importing Libraries & Dataset

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
#plt.style.use("seaborn")

df = pd.read_csv('https://raw.githubusercontent.com/GUC-DM/W2025/refs/heads/main/data/loan_data.csv')
df.head()

ApplicationNumber  Age AnnualIncome  CreditScore EmploymentStatus  \
0              26231   18   $26,992.00          586         Employed   
1              13176   18   $22,825.00          581    Self-Employed   
2              27751   18   $75,891.00          522         Employed   
3              20628   18   $92,094.00          565         Employed   
4              28773   18   $35,686.00          554         Employed   

  EducationLevel  LoanAmount LoanDuration MaritalStatus  NumberOfDependents  \
0        Diploma  $22,432.00    36 months        Single                   3   
1         Master  $30,457.00    24 months        Single                   0   
2       Bachelor  $30,583.00    96 months       Married                   2   
3       Bachelor  $72,738.00    36 months        Single                   1   
4        Diploma  $28,021.00    36 months       Married                   3   

  HomeOwnershipStatus  BankruptcyHistory         LoanPurpose  \
0            Mortgage                  0  Debt Consolidation   
1                Rent                  0    Home Improvement   
2            Mortgage                  0            Personal   
3            Mortgage                  0             Vehicle   
4                Rent                  0    Home Improvement   

   PreviousLoanDefaults MonthlyLoanPayment MonthlyIncome  JobTenure  \
0                     0            $623.11     $2,249.33          3   
1                     0          $1,269.04     $1,902.08          5   
2                     0            $318.57     $6,324.25          6   
3                     0                NaN     $7,674.50          3   
4                     0            $778.36     $2,973.83          1   

  LoanApproved  
0           No  
1           No  
2           No  
3           No  
4           No

## Data Inspection

In order to be able to view dataframe summary such as the columns, data types, count of non-null values and memory usage of the dataframe

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ApplicationNumber     20000 non-null  int64 
 1   Age                   20000 non-null  int64 
 2   AnnualIncome          20000 non-null  object
 3   CreditScore           20000 non-null  int64 
 4   EmploymentStatus      16260 non-null  object
 5   EducationLevel        20000 non-null  object
 6   LoanAmount            20000 non-null  object
 7   LoanDuration          20000 non-null  object
 8   MaritalStatus         20000 non-null  object
 9   NumberOfDependents    20000 non-null  int64 
 10  HomeOwnershipStatus   20000 non-null  object
 11  BankruptcyHistory     20000 non-null  int64 
 12  LoanPurpose           20000 non-null  object
 13  PreviousLoanDefaults  20000 non-null  int64 
 14  MonthlyLoanPayment    15660 non-null  object
 15  MonthlyIncome         17160 non-null

* **EmploymentStatus, MonthlyLoanPayment and MonthlyIncome** have missing values
* **AnnualIncome, LoanAmount, LoanDuration, MonthlyLoanPayment,** and **MonthlyIncome** are of type object but they should be numeric data types
* **LoanApproved** should be binary so it will be encoded later on



In [ ]:
pd.set_option("display.float_format", "{:.2f}".format)
df.describe()

ApplicationNumber      Age  CreditScore  NumberOfDependents  \
count           20000.00 20000.00     20000.00            20000.00   
mean            20500.50    39.75       571.61                1.52   
std              5773.65    11.62        51.00                1.39   
min             10501.00    18.00       343.00                0.00   
25%             15500.75    32.00       540.00                0.00   
50%             20500.50    40.00       578.00                1.00   
75%             25500.25    48.00       609.00                2.00   
max             30500.00    80.00       712.00                5.00   

       BankruptcyHistory  PreviousLoanDefaults  JobTenure  
count           20000.00              20000.00   20000.00  
mean                0.00                  0.10       5.00  
std                 0.00                  0.30       2.24  
min                 0.00                  0.00       0.00  
25%                 0.00                  0.00       3.00  
50%                 0.00                  0.00       5.00  
75%                 0.00                  0.00       6.00  
max                 0.00                  1.00      16.00

Check count of missing values for each column

In [ ]:
df.isnull().sum()

ApplicationNumber          0
Age                        0
AnnualIncome               0
CreditScore                0
EmploymentStatus        3740
EducationLevel             0
LoanAmount                 0
LoanDuration               0
MaritalStatus              0
NumberOfDependents         0
HomeOwnershipStatus        0
BankruptcyHistory          0
LoanPurpose                0
PreviousLoanDefaults       0
MonthlyLoanPayment      4340
MonthlyIncome           2840
JobTenure                  0
LoanApproved               0
dtype: int64

* **EmploymentStatus** has 3740 missing values
* **MonthlyLoanPayment** has 4340 missing values
* **MonthlyIncome** has 2840 missing values

**Checking for Unique values in each column**


In [ ]:
for column in df.columns:
    print(f"{column} : {df[column].unique()}")
    print("====================================")

ApplicationNumber : [26231 13176 27751 ... 18749 13396 17392]
Age : [18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41
 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65
 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80]
AnnualIncome : ['$26,992.00' '$22,825.00' '$75,891.00' ... '$111,290.00' '$48,501.00'
 '$95,945.00']
CreditScore : [586 581 522 565 554 531 574 479 621 520 504 435 558 611 505 606 524 588
 557 548 523 594 604 501 585 561 575 551 533 487 539 610 589 525 527 545
 432 568 521 459 507 439 567 602 566 573 583 526 546 515 607 535 597 579
 510 587 550 605 430 486 528 622 580 600 556 458 562 537 598 591 476 603
 564 418 570 559 601 485 467 398 552 494 444 516 497 540 466 433 462 553
 543 576 542 571 599 617 457 475 538 478 596 560 569 584 532 578 625 593
 614 536 518 609 555 547 563 426 636 472 541 474 623 509 513 483 492 496
 620 530 465 517 577 477 508 440 608 493 421 451 616 500 534 613 544 514
 529 572 460 482 506 450 452 519 489 612

**Check number of unique values** for each column

In [ ]:
for column in df.columns:
    print(f"{column}: Number of unique values {df[column].nunique()}")
    print("==========================================================")

ApplicationNumber: Number of unique values 20000
Age: Number of unique values 63
AnnualIncome: Number of unique values 17516
CreditScore: Number of unique values 322
EmploymentStatus: Number of unique values 3
EducationLevel: Number of unique values 5
LoanAmount: Number of unique values 15578
LoanDuration: Number of unique values 10
MaritalStatus: Number of unique values 4
NumberOfDependents: Number of unique values 6
HomeOwnershipStatus: Number of unique values 4
BankruptcyHistory: Number of unique values 1
LoanPurpose: Number of unique values 5
PreviousLoanDefaults: Number of unique values 2
MonthlyLoanPayment: Number of unique values 14195
MonthlyIncome: Number of unique values 15222
JobTenure: Number of unique values 17
LoanApproved: Number of unique values 4


* **ApplicationNumber** has 20000 unique value as it's the identifier which will be dropped later on
* **BankruptcyHistory** has 1 unique value so it's not needed
* **LoanApproved** has 4 unique values ('No' 'Yes' 'no' 'yes') due to inconsistency in formatting which leads to redundancy and will be fixed in data cleaning

In [ ]:
df.drop(['ApplicationNumber','BankruptcyHistory'],axis="columns", inplace=True)

We dropped both ApplicationNumber and BankruptcyHistory
* ApplicationNumber as it's a unique identifier (has a unique instance for each value) so it doesn't add any additional information
* BankruptcyHistory has only 1 unique value so it's the same values for all rows means it's not needed

## Data Cleaning

In [ ]:
df.head()

Age AnnualIncome  CreditScore EmploymentStatus EducationLevel  LoanAmount  \
0   18   $26,992.00          586         Employed        Diploma  $22,432.00   
1   18   $22,825.00          581    Self-Employed         Master  $30,457.00   
2   18   $75,891.00          522         Employed       Bachelor  $30,583.00   
3   18   $92,094.00          565         Employed       Bachelor  $72,738.00   
4   18   $35,686.00          554         Employed        Diploma  $28,021.00   

  LoanDuration MaritalStatus  NumberOfDependents HomeOwnershipStatus  \
0    36 months        Single                   3            Mortgage   
1    24 months        Single                   0                Rent   
2    96 months       Married                   2            Mortgage   
3    36 months        Single                   1            Mortgage   
4    36 months       Married                   3                Rent   

          LoanPurpose  PreviousLoanDefaults MonthlyLoanPayment MonthlyIncome  \
0  Debt Consolidation                     0            $623.11     $2,249.33   
1    Home Improvement                     0          $1,269.04     $1,902.08   
2            Personal                     0            $318.57     $6,324.25   
3             Vehicle                     0                NaN     $7,674.50   
4    Home Improvement                     0            $778.36     $2,973.83   

   JobTenure LoanApproved  
0          3           No  
1          5           No  
2          6           No  
3          3           No  
4          1           No

**Numerical Columns**

In [ ]:
numerical_cols=['Age', 'AnnualIncome','CreditScore','LoanAmount','LoanDuration', 'LoanDuration', 'MonthlyLoanPayment','MonthlyIncome','JobTenure']
df.head()

Age AnnualIncome  CreditScore EmploymentStatus EducationLevel  LoanAmount  \
0   18   $26,992.00          586         Employed        Diploma  $22,432.00   
1   18   $22,825.00          581    Self-Employed         Master  $30,457.00   
2   18   $75,891.00          522         Employed       Bachelor  $30,583.00   
3   18   $92,094.00          565         Employed       Bachelor  $72,738.00   
4   18   $35,686.00          554         Employed        Diploma  $28,021.00   

  LoanDuration MaritalStatus  NumberOfDependents HomeOwnershipStatus  \
0    36 months        Single                   3            Mortgage   
1    24 months        Single                   0                Rent   
2    96 months       Married                   2            Mortgage   
3    36 months        Single                   1            Mortgage   
4    36 months       Married                   3                Rent   

          LoanPurpose  PreviousLoanDefaults MonthlyLoanPayment MonthlyIncome  \
0  Debt Consolidation                     0            $623.11     $2,249.33   
1    Home Improvement                     0          $1,269.04     $1,902.08   
2            Personal                     0            $318.57     $6,324.25   
3             Vehicle                     0                NaN     $7,674.50   
4    Home Improvement                     0            $778.36     $2,973.83   

   JobTenure LoanApproved  
0          3           No  
1          5           No  
2          6           No  
3          3           No  
4          1           No

* **AnnualIncome, LoanAmount, MonthlyLoanPayment, MonthlyIncome** are columns that contain $ and commas so they need to be removed
* **LoandDuration** should also have "months" removed to keep it a numeric column

In [ ]:
df['LoanDuration'] = df['LoanDuration'].str.replace(' months','').astype(int)
money_cols = ['AnnualIncome', 'LoanAmount', 'MonthlyLoanPayment', 'MonthlyIncome']

for col in money_cols:

    df[col] = df[col].replace(r'[\$,]', '', regex=True).astype(float)

df.head()

Age  AnnualIncome  CreditScore EmploymentStatus EducationLevel  LoanAmount  \
0   18      26992.00          586         Employed        Diploma    22432.00   
1   18      22825.00          581    Self-Employed         Master    30457.00   
2   18      75891.00          522         Employed       Bachelor    30583.00   
3   18      92094.00          565         Employed       Bachelor    72738.00   
4   18      35686.00          554         Employed        Diploma    28021.00   

   LoanDuration MaritalStatus  NumberOfDependents HomeOwnershipStatus  \
0            36        Single                   3            Mortgage   
1            24        Single                   0                Rent   
2            96       Married                   2            Mortgage   
3            36        Single                   1            Mortgage   
4            36       Married                   3                Rent   

          LoanPurpose  PreviousLoanDefaults  MonthlyLoanPayment  \
0  Debt Consolidation                     0              623.11   
1    Home Improvement                     0             1269.04   
2            Personal                     0              318.57   
3             Vehicle                     0                 NaN   
4    Home Improvement                     0              778.36   

   MonthlyIncome  JobTenure LoanApproved  
0        2249.33          3           No  
1        1902.08          5           No  
2        6324.25          6           No  
3        7674.50          3           No  
4        2973.83          1           No

**1. Dealing with missing values**
Since **EmploymentStatus** is a categorical attribute, we used "unknown" to replace the missing values

For **MonthlyLoanPayment** and **MonthlyIncome**, we will replace missing values with median since they are both continuous numerical values

In [ ]:
df['EmploymentStatus'] = df['EmploymentStatus'].fillna('Unknown')
df['MonthlyLoanPayment'] = df['MonthlyLoanPayment'].fillna(df['MonthlyLoanPayment'].median())
df['MonthlyIncome'] = df['MonthlyIncome'].fillna(df['MonthlyIncome'].median())
df.head()

Age  AnnualIncome  CreditScore EmploymentStatus EducationLevel  LoanAmount  \
0   18      26992.00          586         Employed        Diploma    22432.00   
1   18      22825.00          581    Self-Employed         Master    30457.00   
2   18      75891.00          522         Employed       Bachelor    30583.00   
3   18      92094.00          565         Employed       Bachelor    72738.00   
4   18      35686.00          554         Employed        Diploma    28021.00   

   LoanDuration MaritalStatus  NumberOfDependents HomeOwnershipStatus  \
0            36        Single                   3            Mortgage   
1            24        Single                   0                Rent   
2            96       Married                   2            Mortgage   
3            36        Single                   1            Mortgage   
4            36       Married                   3                Rent   

          LoanPurpose  PreviousLoanDefaults  MonthlyLoanPayment  \
0  Debt Consolidation                     0              623.11   
1    Home Improvement                     0             1269.04   
2            Personal                     0              318.57   
3             Vehicle                     0              439.19   
4    Home Improvement                     0              778.36   

   MonthlyIncome  JobTenure LoanApproved  
0        2249.33          3           No  
1        1902.08          5           No  
2        6324.25          6           No  
3        7674.50          3           No  
4        2973.83          1           No

Next we want to fix **LoanApproved** column that has 4 unique value instead of 2 (what we mentioned earlier in data inspection)

## Exploratory Data Analysis

**Q1: On average, which type of educational level has the highest approval rate? Show their order on the graph.**

**Visualization**

In [28]:
df['LoanApproved'] = df['LoanApproved'].astype('category').cat.codes
df['LoanApproved']



0        0
1        0
2        0
3        0
4        0
        ..
19995    0
19996    0
19997    0
19998    0
19999    0
Name: LoanApproved, Length: 20000, dtype: int8

In [30]:
order = df[['EducationLevel', 'LoanApproved']].groupby('EducationLevel').mean().sort_values('LoanApproved').index
order

Index(['Bachelor', 'Diploma', 'Doctorate', 'High School', 'Master'], dtype='object', name='EducationLevel')

**Answer for Q1**: Your answer here

First, we encoded the LoanApproved column into binary values (Yes = 1, No = 0), because we cannot calculate a meaningful average on text values. Then we grouped by educationlevel and calculated the mean of LoanApproved computing the approval rate for each group it was found that Bachelor has the lowest loan approval rate and masters has the highest loan approval rate


**Q2: How does the annual income vary among approved applicants? Interpret the values of the 3 quartiles.**

**Visualization**

**Answer for Q2**: Your answer here

**Q3: How does the age of an applicant affect their credit score? (Hint: Use the line of best fit.)**

**Visualization**

**Answer for Q3**: Your answer here

**Q4: Is the distribution of applicants' income per month normal or skewed?**

**Visualization**

**Answer for Q4**: Your answer here

## Data Preparation for Modelling

## Modelling

## Evaluation

## Bonus (Optional)